text/0113-secure-script-container.md

```diff
@@ -0,0 +1,183 @@
+---
+title: Secure Script Container
+status: DRAFTED
+created_at: 2020-07-15
+pr: https://github.com/salesforce/lwc-rfcs/pull/39
+champion: Caridy Patiño, Ted Conn
+---
+
+# Secure Script Container for LWC Apps
+
+## Summary
+
+We need a way for application developers to integrate common third-party libraries into their LWC applications. These libraries, such as Google Analytics, are severely hindered by the shadow DOM which prevents global interaction with any element on the page.
+
+By providing a secure container for scripts to run, see, and interact with all the shadow DOM trees as if it were the light DOM, application developers can continue to run their existing integrations until these libraries support shadow DOM semantics natively.
+
+## Basic example
+
+```html
+<secure-script src="//cdn.optimizely.com/js/12345678.js"></secure-script>
+```
+
+`secure-script` is a custom element which acts the same as the `script` tag, but any code which is evaluated inside this container is able to traverse the entire shadow tree using light DOM semantics. So if the above script runs `document.querySelectorAll('button')`, it will return all button elements regardless if they are in the shadow or not.
+
+```html
+<secure-script>
```

Comment:
I would love to see the equivalent imperative inline example, something like:

```js
SecureScript.evaluate(() => {});
```

Reply:
That's a very fair point, we can do a follow up, however I would love to mention it here somewhere.

Reply:
yes, we can mention it. I believe a static method on the VirtualScript web component class is sufficient, so you could import the class, and register yourself, extend it, or just call the static evaluator method. I will add a note for that to be a follow up proposal.